### ISSUE TITLE
Javascript not loaded in time for invokeAsync call

### ISSUE THREAD

**Issue by limefrogyank** [SQL-MisterMagoo/BlazorEmbedLibrary#4]
Perhaps I am not using your library correctly, but I put the `EmbeddedContent` tag in my top-most view of a Razor Components (I guess it's back to Blazor Server-side) app.  

```
<EmbeddedContent BaseType="@(typeof(Component1))" />
```
However, in the `OnInitAsync` overload for one of my pages that loads pretty much immediately, I eventually make `IJSRuntime.InvokeAsync` call to some javascript in one of the embedded files.  I get an exception that my javascript object can't be found.  I took my same component file and hard-coded it into the main app inside the Index.cshtml file and it works fine.  

Is there a limitation to this library for loading javascript files?  Thanks!
(FYI, I love your IsPreRendering component.)

**Comment by Simon-Miller** [comment]
I'm struggling with the library to - and hope we get this in Blazor's core functionality soon.  But I suspect we'd hit the same issue, if I'm reading you correctly. Signal R aspect of the startup.  Keep in mind that your javascript needs to load, and the SignalR needs to load too.  SignalR should load AFTER your code.  So yours should appear before: 
<script src="_framework/blazor.server.js"></script>

You're lucky if it's generating a script request!  I can't seem to figure out where I'm going wrong, and have gone down the road of adding the middleware to the startup and telling it about the assembly, and then in the main html structure (for me, _Host.cshtml) where the html and body tags are located, and also the script request for _framework/blazor.server.js - - I'm adding a script line above that asking for the specific filename (no path) that my component needs. 

Basically, I've followed the example code in the GitHub repository.

**Comment by SQL-MisterMagoo** [comment]
If you can supply some sample code I'll take a look and see what's wrong.

**Comment by Simon-Miller** [comment]
Thank you kindly.
It's pretty much the default server-side blazor template, and adding the default blazor component.
Then I created a new blazor component to actually play with, and develop a canvas control interop.  yeah, there's a project out there for that already - but what would I learn if I just use someone else's?

To sumarize then, I believe I followed your instructions.  I added the TagHelper, I added a @using for your helper - and I tried adding the <EmbeddedContent BaseType="@typeof(Canvas2D)"></EmbeddedContent> right after I reference the component <Canvas2D ... />

As that didn't work, I thought I'd try the function that creates a list of assemblies, and reference that instead, like your example code.  That didn't seem to do anything either.
I tried moving the <EmbeddedContent>  to the _Host file, which is the html and body tags, along with the Blazor client-side code.
I didn't see any output in the HTML from this either.
Which is why I ended up following your example and adding your code as middleware.
I noted that the it seemed to pick up the name of the file without a path.  I don't know if it works with the path or not - but the code in the zip references the file by name without any path, and works.

I hope that gives you an idea of my thinking and approach.
Really appreciate any feedback -including the bit where you tell me what an idiot I've been!

[BlazorComponents3.zip](https://github.com/SQL-MisterMagoo/BlazorEmbedLibrary/files/3229497/BlazorComponents3.zip)

**Comment by SQL-MisterMagoo** [comment]
@Simon-Miller There were a few issues - please find a working version of your solution attached - search for "MMFIX" to find where I added code. You had added BlazorEmbedLibrary to your component library, which is incorrect usage, so I removed it.

[BlazorComponents3.zip](https://github.com/SQL-MisterMagoo/BlazorEmbedLibrary/files/3230166/BlazorComponents3.zip)

@limefrogyank Sorry I didn't see your issue before.

@All It seems Server side Blazor has changed a bit since I wrote this - I'll update the instructions in the README shortly.

TLDR; Pre-rendering means we can't use JSInterop until Blazor tells us we are connected - IComponentContext helps here with "IsConnected" - but that does not refresh if you only reference it in markup.... so you need a bit of helper code in OnAfterRender to give it a nudge....

**Comment by Simon-Miller** [comment]
Ah, yes, glad you mentioned that for anyone reading.  I forgot to mention that.  I was wondering if having the component depend upon your library, and thus be able to add the razor  <EmbeddedContent ... /> would work.  If it can work? (I don't see why not?) 

This would give a better separation of concerns.
I'm thinking that if you provide a component base class of your own - and have new blazor components inherit from it - then in the startup you could scan for classes in assemblies that inherit from your base class?  (or interface).  That would remove the need to register all the components.  I did something similar a number of years ago - but I found it hard to guarantee that the assembly was already loaded, and it would sometimes miss classes.  The issue was self-healing when I brought dependency injection in, as that all sets up first, and the process of registering all the classes with a dependency injeciton container meant the assemblies would need to be loaded in order to register its types.  
Might be worth a thought?

I've tried your fixes out.  Thanks so much!  If I've any improvements from playing around, I'll let you (and anyone listening) know here.

**Comment by Simon-Miller** [comment]
I've managed to scan all the loaded assemblies for components, and handed them off to your code:

```
<EmbeddedContent Assemblies=@ComponentLibraryAssemblies />
@functions
{
    private static List<Assembly> _componentLibraryAssemblies = null;
    private List<Assembly> ComponentLibraryAssemblies
    {
        get
        {
            if (_componentLibraryAssemblies == null)
            {
                var allAssemblies = AppDomain.CurrentDomain.GetAssemblies();

                var assembliesToScan = allAssemblies.Where(asm => asm.GetTypes()
                    .Where(t => t.IsClass && t.IsSubclassOf(typeof(ComponentLibBase))).Count() > 0)
                    .ToList();

                _componentLibraryAssemblies = assembliesToScan;
            }
            
            return _componentLibraryAssemblies;
        }
    }
}
```

But I can't find a way of controlling a component's rendering from a base-class.  The BuildRenderTree(x) doesn't seem to get called?  It talks about using the IComponent interface instead of the ComponentBase class in intellise.  But I can't see any documented examples around this.
So anyway - your code still wins!  But I thought you might be interested in the assembly scanning.  It might be painfuil on first page load, but because it keeps the results in a static property, subsequent calls to the server should be much improved.  If you need performance from first-page-load, then your solution is fine, but places the responsibility on the developer to manually register all types.

Oh, in case you (the reader) and wondering, I'm trying to give a component library a base class to take care of the concerns that @SQL-MisterMagoo addressed in his last post.  His code delays rendering a component by checking if the page is fully loaded before forcing a refresh, which will then see the page is fully loaded, and render the component.  This is an attempt to delay code execution, giving the embedded scripts a chance to load in the client before your component starts acting upon it.  Corrent me if I'm unclear or simply wrong.

**Comment by SQL-MisterMagoo** [comment]
@Simon-Miller Thanks for taking the time to share that information. 

I'm currently waiting to see what Preview 6 brings as it will hopefully make this package redundant!

If it turns out that there is still a need for more control after the official handler is available, I will add some dependency scanning like you have shown.
